petsc: 3.8.4 -> 3.10.3

jamtrott opened the pull request:
These changes upgrade PETSc to the most recent release and add several useful linear solver backends, including HYPRE, MUMPS, scalapack, and SPAI.  In addition, the Python wrapper package petsc4py is built.

This has been tested on Mac OS X 10.13.6 (High Sierra).

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

jamtrott committed:
scotch: Enable building on darwin

Also, build libesmumps and libptesmumps to later support the mumps
package.

jamtrott committed:
petsc: 3.8.4 -> 3.10.3

jamtrott committed:
pythonPackages.petsc4py: init at 3.10.1

jamtrott committed:
hypre: init at 2.14.0

jamtrott committed:
mumps: init at 5.1.2

jamtrott committed:
scalapack: Enable building on darwin

jamtrott committed:
spai: init at 3.0

jamtrott committed:
pythonPackages.mpi4py: Re-compile cython code

This prevents issues when using multiple cython-based packages
together (for example, mpi4py and petsc4py) due to code that has been
generated with incompatible cython versions.

Also, update homepage.

jamtrott committed:
maintainers: add jamtrott (James D. Trotter)

srhb commented on pkgs/development/libraries/science/math/hypre/default.nix:
```diff
@@ -0,0 +1,27 @@
+    homepage = https://computation.llnl.gov/projects/hypre-scalable-linear-solvers-multigrid-methods;
+    license = stdenv.lib.licenses.lgpl21;
+    platforms = stdenv.lib.platforms.all;
+    maintainers = with lib.maintainers; [ jamtrott ];
```
`stdenv.lib`

srhb commented on pkgs/development/libraries/science/math/mumps/default.nix:
```diff
@@ -0,0 +1,41 @@
+    homepage = http://mumps-solver.org/;
+    license = stdenv.lib.licenses.cecill-c;
+    platforms = stdenv.lib.platforms.all;
+    maintainers = with lib.maintainers; [ jamtrott ];
```
`stdenv.lib`

srhb commented on pkgs/development/libraries/science/math/petsc/default.nix:
```diff
@@ -1,50 +1,90 @@
     homepage = https://www.mcs.anl.gov/petsc/index.html;
     platforms = stdenv.lib.platforms.all;
     license = stdenv.lib.licenses.bsd2;
+    maintainers = with lib.maintainers; [ jamtrott ];
```
`stdenv.lib`

srhb commented on pkgs/development/libraries/science/math/spai/default.nix:
```diff
@@ -0,0 +1,35 @@
+    description = "Sparse approximate inverse preconditioner";
+    homepage = https://bitbucket.org/petsc/pkg-spai;
+    platforms = stdenv.lib.platforms.all;
+    license = lib.licenses.unfree;
```
`stdenv.lib` for both license and maintainers.

srhb commented on pkgs/development/python-modules/petsc4py/default.nix:
```diff
@@ -0,0 +1,43 @@
+    description = "Python bindings for PETSc, the Portable, Extensible Toolkit for Scientific Computation";
+    homepage = https://bitbucket.org/petsc/petsc4py;
+    license = stdenv.lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ jamtrott ];
```
`stdenv.lib`

srhb reviewed (changes requested):
Evaluation currently fails due to some misses in the meta sections. Can you fix these up? :)

jamtrott commented:
> Evaluation currently fails due to some misses in the meta sections. Can you fix these up? :)

Hi Sarah. Thank you for reviewing!

I fixed those mistakes and have pushed some new commits. I'm not sure whether it was necessary to create a commit for each separate package, but they can always be squashed before merging anyway.

srhb commented:
Ideally I'd like this to be squashed down to one commit per package, would you mind doing that? :)

jamtrott commented:
> Ideally I'd like this to be squashed down to one commit per package, would you mind doing that? :)

No problem at all. What would be the correct way of doing so? Do I close this pull request and submit a new one? Or is it possible to overwrite the branch with a forced push?

srhb commented:
@jamtrott The easiest way is the locally undo your changes with a soft reset, eg `git reset master` -- this will leave your changed files but uncommitted. You can now piecemeal commit each logical package, and finally force push your branch, which will update this PR. :)

dotlambda commented on pkgs/development/python-modules/mpi4py/default.nix:
```diff
@@ -46,14 +46,14 @@ buildPythonPackage rec {
   '';
 
-  setupPyBuildFlags = ["--mpicc=${mpi}/bin/mpicc"];
+  setupPyBuildFlags = [ "--mpicc=${mpi}/bin/mpicc" "build_src --force" ];
```
Can you explain what `build_src --force does`? Maybe a comment inside the expression would be nice.

dotlambda commented on pkgs/development/python-modules/petsc4py/default.nix:
```diff
@@ -0,0 +1,43 @@
+  meta = {
+    description = "Python bindings for PETSc, the Portable, Extensible Toolkit for Scientific Computation";
+    homepage = https://bitbucket.org/petsc/petsc4py;
+    license = stdenv.lib.licenses.bsd3;
```
```suggestion
    license = lib.licenses.bsd3;
```

dotlambda commented on pkgs/development/python-modules/petsc4py/default.nix:
```diff
@@ -0,0 +1,43 @@
+{ stdenv
```
```suggestion
{ lib
```

dotlambda commented on pkgs/development/python-modules/petsc4py/default.nix:
```diff
@@ -0,0 +1,43 @@
+    openmpi
+    openssh
+    mpi4py
+  ];
```
Some of those should probably be propagated. Why is python there?

jamtrott replied:
Hi @dotlambda, thank you for your comments. I admit that I'm not sure about buildInputs versus propagatedBuildInputs. Do you have any suggestions or guidelines for determining which packages should be propagated?

The python package is not needed, and I have removed it from the list.

dotlambda replied:
See https://nixos.org/nixpkgs/manual/#reference:

> If something is exclusively a build-time dependency, use buildInputs; if it is (also) a runtime dependency, use propagatedBuildInputs.

dotlambda replied:
See above for `cython`.
I don't find `mpi4py` in their [setup.py](https://bitbucket.org/petsc/petsc4py/src/6fe2748c55f524f51f93dc8d119d650fee44f9d0/setup.py).
`numpy` is in `install_requires` and should thus be propagated.
What do you infer the dependency on openssh from?

dotlambda reviewed (commented):
Please add the `pythonPackages.` prefix to the respective commits.

jamtrott commented:
@srhb, @dotlambda, would you agree that the latest revisions address the issues you have raised?

dotlambda commented on pkgs/development/python-modules/petsc4py/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+  PETSC_DIR = "${petsc}";
+
+  setupPyBuildFlags = [ "build_src --force" ];
```
You should add the same comment here.

jamtrott replied:
Done.

dotlambda commented on pkgs/development/python-modules/petsc4py/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+  propagatedBuildInputs = [
+    petsc
+    openmpi
```
I don't find `openmpi` in ther setup.py.
`petsc` can probably be dropped from here since you specify `PETSC_DIR`.

jamtrott replied:
I moved `openmpi` to `nativeBuildInputs` and removed `petsc`.  The package doesn't build without `openmpi`.

jamtrott committed:
pythonPackages.mpi4py: Move cython to nativeBuildInputs

jamtrott committed:
pythonPackages.petsc4py: Fix build inputs and add missing comment

Remove mpi4py and make other packages nativeBuildInputs,
except numpy, which is propagated.

jamtrott commented on pkgs/development/python-modules/mpi4py/default.nix:
```diff
@@ -46,14 +46,18 @@ buildPythonPackage rec {
+  setupPyBuildFlags = [ "--mpicc=${mpi}/bin/mpicc" "build_src --force" ];
 
-  buildInputs = [ mpi openssh ];
+  buildInputs = [ cython mpi openssh ];
```
Thanks, this should be fixed now.

tkonolige commented:
Would it be possible to parametrize these packages by which MPI implementation they use instead of hardcoding openmpi?

FRidh commented:
Would be good to see the merge conflicts resolved, because this seems otherwise ready.
